### PR TITLE
Lazily build inherited lookup maps, but then eagerly build them as so…

### PR DESCRIPTION
**Pull request by dnfield**
See also https://github.com/flutter/flutter/pull/95596 which got reverted in https://github.com/flutter/flutter/pull/100152

This is a slightly different version of that change that improves worst frame build times.

The thrust of this change is that it avoids building the inherited lookup type maps until they're actually needed. Based on experimentation in new gallery, there are a number of inherited widgets that never get used for a lookup. Currently we build maps for them that never get used. This patch avoids that.

It still has worst case quadratic space.  Here are some before and after results from a low end Android device:

BEFORE - Wembley
```
    "average_frame_build_time_millis": 4.425031804733731,
    "90th_percentile_frame_build_time_millis": 8.439,
    "99th_percentile_frame_build_time_millis": 42.163,
    "worst_frame_build_time_millis": 249.196,
    "missed_frame_build_budget_count": 71,
    "average_frame_rasterizer_time_millis": 15.346655580192166,
    "90th_percentile_frame_rasterizer_time_millis": 26.531,
    "99th_percentile_frame_rasterizer_time_millis": 70.397,
    "worst_frame_rasterizer_time_millis": 876.068,
    "missed_frame_rasterizer_budget_count": 349,
    "frame_count": 1352,
    "frame_rasterizer_count": 1353,
```

AFTER - Wembley
```
    "average_frame_build_time_millis": 4.324401315789478,
    "90th_percentile_frame_build_time_millis": 8.558,
    "99th_percentile_frame_build_time_millis": 41.234,
    "worst_frame_build_time_millis": 241.986,
    "missed_frame_build_budget_count": 76,
    "average_frame_rasterizer_time_millis": 15.396707815924024,
    "90th_percentile_frame_rasterizer_time_millis": 26.941,
    "99th_percentile_frame_rasterizer_time_millis": 79.116,
    "worst_frame_rasterizer_time_millis": 432.783,
    "missed_frame_rasterizer_budget_count": 373,
    "frame_count": 1368,
    "frame_rasterizer_count": 1369,
```

(shorter benchmark run)
BEFORE - iPhone 11
```
  "average_frame_build_time_millis": 0.742842857142857,
  "90th_percentile_frame_build_time_millis": 0.89,
  "99th_percentile_frame_build_time_millis": 0.942,
  "worst_frame_build_time_millis": 1.248,
  "missed_frame_build_budget_count": 0,
  "average_frame_rasterizer_time_millis": 2.7787571428571423,
  "90th_percentile_frame_rasterizer_time_millis": 2.928,
  "99th_percentile_frame_rasterizer_time_millis": 3.032,
  "worst_frame_rasterizer_time_millis": 3.146,
  "missed_frame_rasterizer_budget_count": 0,
  "frame_count": 70,
  "frame_rasterizer_count": 70,
```

AFTER - iPhone 11
```
  "average_frame_build_time_millis": 0.7151449275362319,
  "90th_percentile_frame_build_time_millis": 0.855,
  "99th_percentile_frame_build_time_millis": 0.916,
  "worst_frame_build_time_millis": 1.082,
  "missed_frame_build_budget_count": 0,
  "average_frame_rasterizer_time_millis": 2.7357142857142867,
  "90th_percentile_frame_rasterizer_time_millis": 2.913,
  "99th_percentile_frame_rasterizer_time_millis": 3.147,
  "worst_frame_rasterizer_time_millis": 3.195,
  "missed_frame_rasterizer_budget_count": 0,
  "frame_count": 69,
  "frame_rasterizer_count": 70,
```

I've kicked off one of customer: money's tests with this patch as well and will post some results from that when they are ready.

I think this should get a test exemption as a refactor with no semantic change, and something that will get monitored by performance benchmarks. The tests to make sure lookups still work all pass.

For new gallery on iPhone, this does not seem to meaningfully move memory numbers.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -5238,16 +5238,33 @@ class InheritedElement extends ProxyElement {
...
-  void _updateInheritance() {
-    assert(_lifecycleState == _ElementLifecycle.active);
-    final Map<Type, InheritedElement>? incomingWidgets = _parent?._inheritedWidgets;
-    if (incomingWidgets != null)
-      _inheritedWidgets = HashMap<Type, InheritedElement>.of(incomingWidgets);
-    else
-      _inheritedWidgets = HashMap<Type, InheritedElement>();
-    _inheritedWidgets![widget.runtimeType] = this;
+  void deactivate() {
+    _lookupTable.clear();
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/framework.dart:5266]: In local tests, `if (_lookupTable.isEmpty) print('${widget.runtimeType} was not touched');` fires more than I would have expected for new gallery.

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: question: does this happen for inherited elements that are FocuNodes or ActionMarkers?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -5238,16 +5238,33 @@ class InheritedElement extends ProxyElement {
...
+  void _updateLookupTable() {
+    if (_lookupTable.isNotEmpty) {
+      return;
+    }
+    _lookupTable[widget.runtimeType] = this;
+    InheritedElement? ancestor = _inheritedAncestor;
+    while (ancestor != null) {
+      // Reading out the ancestor's lookup table is slower than just running
+      // up the tree based on local experiments.
+      _lookupTable.putIfAbsent(ancestor.widget.runtimeType, () => ancestor!);
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/framework.dart:5253]: how do you avoid overrwriting types that are provided more than once?

Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:5253]: putIfAbsent means I'm taking the first one I find as I traverse up the tree.

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5253]: oh I see!

`_lookupTable[ancestor.widget.runtimeType] ??= ancestor!`

Will be marginally faster, especially on web

**Comment by dnfield** [comment]
It appears I have broken something here.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: If I add a print here, run gallery, and tap material, I see:

```
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
```

Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: Tapping around and scrolling a list:

```
flutter: DefaultTextStyle
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
flutter: DefaultTextStyle
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: IconTheme
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: IconTheme
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
flutter: _ParentInkResponseProvider
flutter: _ActionsMarker
flutter: _FocusMarker
flutter: MediaQuery
flutter: IconTheme
flutter: DefaultTextStyle
```

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -4245,7 +4243,9 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
 
   void _updateInheritance() {
     assert(_lifecycleState == _ElementLifecycle.active);
-    _inheritedWidgets = _parent?._inheritedWidgets;
+    _inheritedAncestor = _parent is InheritedElement?
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/framework.dart:4246]: You can avoid a type check by adding a private getter to Element and InheritedElement.

Element.getInheritedAncestor returns the value of the field
InheritedElement.getInheritedAncestor return itself

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: This was wrong. The actions marker and focus markers are often looked up from what I can tell. When I properly guard the print the actions marker only shows up once as not getting looked up.

Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: _ParentInkResponseProvider on the other hand shows up frequently as not accessed.

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: I wonder if we could get a similar effect by allowing only certain inherited elements to be lazy. _ActionsMarker/_FocusMarker won't be used at all if the user isn't interacting with the focus system or a kb

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: ahh OK. Still, that is a lot of inherited widget copying for a a single type of widget. I still think the partitioning idea might be onto something...

File: packages/flutter/lib/src/widgets/framework.dart (hunk 3 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:4246]: How would I set the value of the field though?

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: Clearing the list here is causing problems for FocusScope, which wants to do a lookup when it unfocuses...?

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart:5266]: its an inherited notifer

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart]: Going with it.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 3 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/framework.dart]: Added a polymorphic getter for `_thisOrParentAsInherited`, which `InheritedElement` overrides to return `this`.

**Comment by gaaclarke** [comment]
Are you ready for this to be reviewed now?  Looks like the CI is passing.  I was waiting around for those in case there were bugs.

**Comment by dnfield** [comment]
Should be good for review. Unfortunately in local runs on my wembley this is now a bit more noisy although still seems better on average. I'm trying some runs on a Moto G4 and on an internal benchmark for customer: money as well.

**Comment by gaaclarke** [comment]
I'm a bit wary of a lazy solution that doesn't have have performance savings that are outside the margin of error.  Looks like we are seeing 2% - 3% savings?  It's notoriously difficult to reason about the implications of making something Lazy.  It would be nice if we could say we have at least a 50% probability that this is faster for `customer:money`.

This change seems to optimize the calculation of `_inheritedWidgets`, which I haven't seen strong evidence that it takes a significant amount of time.  In `customer:money` I've seen `_updateInheritance` only take up 5ms.  Maybe it would help if you recapped all your evidence that this will be helpful.  I've seen so much now I might be mixed up.

**Comment by dnfield** [comment]
I am no longer convinced this is helping. I can't seem to replicate the improvement in any non-noisy way, and the current setup at head is performing consistently better.

@gaaclarke: My goal here was to avoid having as many map copies (HashMap.of) and avoid having hash maps for elements that never end up using them. I think I achieved that, but it's not really showing better results unfortunately.

**Pull request closed without merging**